tardis/montecarlo/tests/test_cmontecarlo.py

```diff
@@ -1,99 +1,343 @@
-            ('next_line_id', c_int),
-            ('current_shell_id', c_int),
-            ]
+"""
+Unit tests for methods in `tardis/montecarlo/src/cmontecarlo.c`.
+* `ctypes` library is used to wrap C methods and expose them to python.
+
+
+Probable Reasons for Failing Tests:
+-----------------------------------
+
+1. Change made in C struct declarations:
+  - Reflect the changes done in C structs, into Python counterparts.
+  - Check **tardis/montecarlo/struct.py**.
+
+2. Return type of any method changed:
+  - Modify the `restype` parameter in the test method here.
+  - For example:
+        ```
+        cmontecarlo_methods.rpacket_doppler_factor.restype = c_double
+        ```
+
+3. Underlying logic modified:
+  - Check whether the changes made in C method are logically correct.
+  - If the changes made were correct and necessary, update the corresponding
+    test case.
+
+
+General Test Design Procedure:
+------------------------------
+
+Please follow this design procedure while adding a new test:
 
+1. Parametrization as per desire of code coverage.
+  - C tests have different flows controlled by conditional statements.
+    Parameters checked in conditions can be provided in different testcases.
+  - Keep consistency with variable names as (in order):
+    - `packet_params`
+    - `model_params`
+    - `expected_params` (`expected` if only one value to be asserted.)
+
+2. Test Method body:
+  - Keep name as `test_` + `(name of C method)`.
+  - Refer to method `test_rpacket_doppler_factor` below for description.
+"""
+
+import os
 import pytest
-import numpy as np
+from ctypes import CDLL, byref, c_uint, c_int64, c_double, c_ulong
 from numpy.testing import assert_almost_equal
 
-#pytestmark = pytest.mark.skipif(True, reason='problem with the files')
-
 from tardis import __path__ as path
+from tardis.montecarlo.struct import RPacket, StorageModel, RKState
 
+# Wrap the shared object containing tests for C methods, written in C.
+# TODO: Shift all tests here in Python and completely remove this test design.
 test_path = os.path.join(path[0], 'montecarlo', 'test_montecarlo.so')
+cmontecarlo_tests = CDLL(test_path)
+
+# Wrap the shared object containing C methods, which are tested here.
+cmontecarlo_filepath = os.path.join(path[0], 'montecarlo', 'montecarlo.so')
+cmontecarlo_methods = CDLL(cmontecarlo_filepath)
+
+
+@pytest.fixture(scope="function")
+def packet():
+    """Fixture to return `RPacket` object with default params initialized."""
+    packet_default = {
+        'nu': 0.4,
+        'mu': 0.3,
+        'energy': 0.9,
+        'r': 7.5e14,
+        'tau_event': 2.9e13,
+        'nu_line': 0.2,
+        'current_shell_id': 0,
+        'next_line_id': 1,
+        'last_line': 0,
+        'close_line': 0,
+        'current_continuum_id': 1,
+        'virtual_packet_flag': 1,
+        'virtual_packet': 0,
+        'status': 0,
+        'id': 0
+    }
+    return RPacket(**packet_default)
+
+
+@pytest.fixture(scope="function")
```

Comment:
`TypeError: byref() argument must be a ctypes instance, not 'function'`  
Upon using `autouse`, this pops out and doesn't rectify, I tried some solutions but nothing works.